synapse/storage/util/id_generators.py

```diff
@@ -210,6 +211,21 @@ def __init__(
         # should be less than the minimum of this set (if not empty).
         self._unfinished_ids = set()  # type: Set[int]
 
+        # We track the max position where we know everything before has been
+        # persisted. This is done by a) looking at the min across all instances
+        # and b) noting that if we have seen a run of persisted positions
+        # without gaps (e.g. 5, 6, 7) then we can skip forward (e.g. to 7).
+        #
+        # Note: There is no guarentee that the IDs generated by the sequence
+        # will be gapless; gaps can form when e.g. a transaction was rolled
+        # back. This means that sometimes we won't be able to skip forward the
+        # position even though everything has been persisted. However, since
+        # gaps should be relatively rare it's still worth while doing this.
```

Comment:
I'm not sure what this last sentence means. Since gaps should be relatively rare it's still worth doing...what?

Reply:
It does feel like I got bored and just gave up at this point

Reply:
Let me know if that's better? I feel like its still a bit poorly constructed :/

Reply:
It seems OK. I almost wonder if this should go below where the gap checking is, but I think it is fine the way it is.